Feature / Achievements, holes, portals, finish-args, search, changelog

Arxcis opened the pull request:
Fixes #3

### Functional review help
1. Run python3 server
```
make server
```
2. Open browser at `0.0.0.0:8000/`
3. Click around

### Demo - Sort by holes

![image](https://user-images.githubusercontent.com/3092618/99327714-d2a3bf80-287a-11eb-8391-4adefe54b8a5.png)

Arxcis committed:
Add new /achievements page

Arxcis committed:
Rename search -> list and achievements -> spec

Arxcis committed:
List 5 achievements and 4 holes

Arxcis committed:
Use the new counting system based on achievements and holes

Arxcis committed:
Rename back to achievements again

Arxcis committed:
Implemented achievements and hole lists

Arxcis committed:
Cleaner title on list rows

Arxcis commented:
@tchx84 Maybe you want to review this PR? I would be grateful :pray:

Arxcis committed:
Rename /list -> /lists

Arxcis committed:
More graph title

Arxcis committed:
Refactor src/ -> lib/ and cmd/

Arxcis committed:
Add source: github.com/flathub to all the charts

Arxcis committed:
Increase width from 700px -> 830px

Arxcis committed:
Remove unused _config.yml

Arxcis committed:
Better text on achievements and holes

Arxcis committed:
Add title= to achievement and hole articles

Arxcis committed:
Use the newest achievements and holes in the graphs

Arxcis committed:
Make xdg-run a special edge-case

Arxcis committed:
Remove redundant is not xdg-run check

Arxcis committed:
fix typo "to an entire filesystem"

Arxcis committed:
Show medal on achievements in all graphs

Arxcis committed:
Fix bug, which discovered a lot more --filesystem-all holes

Arxcis committed:
an -> the

Arxcis committed:
Rename lib/achievemnts -> lib/counting.js

Arxcis committed:
Sum holes

Arxcis committed:
Rename counting.js -> stats.js

Arxcis committed:
Better comments in stats.js

Arxcis committed:
better var names in sum functions in stats.js

Arxcis committed:
Rename talkNames -> portals

Arxcis committed:
Add portal graphs

Arxcis committed:
Fix crash on /lists

Arxcis committed:
Rename /achievements/ -> /library/ and add #portals section

Arxcis committed:
Add portals opened to totals graph

Arxcis committed:
Add portals list

Arxcis committed:
remove the medal from --no-device and --no-pulseaudio

- As it is not clear to me that this is actually an
achievement

Arxcis committed:
Fix missing total portals

Arxcis committed:
Display 24 portals! :)

Arxcis committed:
Don't show items in list with zero star count or portal count or ...

Arxcis committed:
Display ellipsis list

Arxcis committed:
Hover on child when parent-hover

Arxcis committed:
Show 1-20 portals

Arxcis committed:
Put #number in front of all portal labels

tchx84 commented:
Ill wait for this PR to converge, since it's still changing rapidly, but I have one suggestion, don't mix data and code. If you generate data, just drop it in a json file and load it, but don't mix it with the code.

Arxcis commented:
@tchx84 Which data are you specifically talking about. Do you mean the `db/flathub/*.js` files?

I wanted to dump them as `.json` files initially, but the browser ES modules does not accept `.json` as mimetype. At least when I tried in latest Firefox:

**Does not work**
```js
import count from "./data/flathub/count.json"
```
**Works**
```js
import count from "./data/flathub/count.js"
```
I tried to separate the data from the code, by putting in a separate `data`-folder.

**Update**
- Renamed the `db/`-folder -> `data/` in 6c802fc to make it clearer that it is data-only.
- Added `## Folder structure`-section to the README.md in e150fea

Arxcis committed:
Rename db/-folder -> data/

Arxcis committed:
README.md / Add ## Folder structure

tchx84 commented:
> @tchx84 Which data are you specifically talking about. Do you mean the `db/flathub/*.js` files?

Yes, the proper way of doing this is: https://stackoverflow.com/questions/16991341/json-parse-file-path

Arxcis commented:
> Yes, the proper way of doing this is: https://stackoverflow.com/questions/16991341/json-parse-file-path

Ah, of course! Thanks :+1: So in modern javascript that would look like:
```js
const res = await fetch("./data/flathub/count.json")
const count = await res.json()
```

Arxcis committed:
renamed the data/ files from .js -> json

Arxcis committed:
Rename repos -> apps

Arxcis committed:
Add share-alike license badge to about page

Arxcis committed:
Run make makefiles and make count to make sure they output correct json

Arxcis committed:
Convert app to load data as .json instead of .js

Arxcis commented on index.html:
```diff
@@ -113,241 +113,247 @@ <h1 id="title" title="Flatstat | Graph">flatstat</h1>
-    },
-  });
+  async function renderCharts() {
+    const count = await useJSON("./data/flathub/count.json")
```
**Update:** Loading data as `.json` instead of `.js` here...

Arxcis committed:
Remove console.logs

Arxcis commented on lists/index.html:
```diff
@@ -34,36 +34,20 @@ <h1 id="title" title="Flatstat | List">flatstat</h1>
-  })
+  // Map, filter, sort, slice, join list
+  async function renderList() {
+    const metafiles = await useJSON("../data/flathub/metafiles.json")
```
....and here

Arxcis committed:
Update README

Arxcis committed:
stats.js / add newlines between mini-functions for readability

Arxcis committed:
fix missing env value bug

Arxcis committed:
Add xdg-cache and xdg-data to well-known filesystem

Arxcis committed:
Renamed cmd/*.js -< cmd/make*.js

Arxcis committed:
Deduplicate metafiles.json

Arxcis committed:
Make count work with the new metadata.json structure

Arxcis committed:
Fix missing --no-device and --no-filesystem

Arxcis committed:
Count finishArgs

Arxcis committed:
Add list of --finish-args and search!

Arxcis committed:
Move /lists -> / and / -> /graphs

Arxcis committed:
Make row unclickable when no displayURL is present

Arxcis committed:
Make search value persist between tabs

Arxcis commented:
## Demo: Search and finish-args

<img src="https://user-images.githubusercontent.com/3092618/99727250-fa339b80-2ab7-11eb-9d29-fa819486ebf9.png" width=600>

Arxcis committed:
--finish-arg - Used by n apps

Arxcis committed:
Remove count suffix from list titles

Arxcis committed:
Add 1px top margin

Arxcis committed:
Show "No results" when there are no search results

Arxcis committed:
Add support for ?search=

Arxcis committed:
Group finish-args with one occurance

- if they starsWith() finish arg with more than one occurance.
- Group them into an <other> group

Arxcis committed:
Support multiple search terms

Arxcis committed:
Support " and space in url

Arxcis committed:
Allow search on multiple columns of data

Arxcis committed:
Hide numbers from seach result details

Arxcis committed:
Merge #apps and #stars

tchx84 commented:
Looking great so far :+1:

Arxcis commented:
@tchx84 Thanks! It was a good idea of you to wait with the review :sweat_smile: Things are changing a lot now in the early phase. Looking forward when things will stabilize :pray: I think I am getting close.

Arxcis committed:
Add changelog

Arxcis commented:
## Demo changelog
<img src="https://user-images.githubusercontent.com/3092618/99740016-5dc8c380-2ace-11eb-8f24-559c18e7d772.png" width=600>
<img src="https://user-images.githubusercontent.com/3092618/99740194-bef09700-2ace-11eb-99d9-7b7c276f3fbe.png" width=600>

Arxcis commented:
Ok, now I have to stop adding more features. This PR is getting too big.. :sweat_smile:

Arxcis committed:
Include additions from beginning of time in changelog

Arxcis committed:
Make changelog fast again

Arxcis committed:
Update data to 2020-11-19

Arxcis committed:
Improve search performance to render entire changelog history

Arxcis commented:
**Get a list of the new apps this month**
<img src="https://user-images.githubusercontent.com/3092618/99845198-8905ee00-2b74-11eb-8197-42f12d9eeff3.png" width=600>
